JASP-Desktop/analysis/analysisform.cpp

```diff
@@ -721,49 +692,55 @@ void AnalysisForm::clearControlError(JASPControlBase* control)
 
 void AnalysisForm::clearFormErrors()
 {
-	if (Utils::currentSeconds() - _lastAddedErrorTimestamp > 5)
-	{
-		_formErrorMessages.clear();
-		_setErrorMessages();
+	_formErrors.clear();
+	emit errorsChanged();
+}
 
-		// Remove also warning without text
-		QSet<JASPControlBase*> controlsWithWarning = _jaspControlsWithWarningSet; // Copy the set: the setHasWarning method changes it.
 
-		for (JASPControlBase* control : controlsWithWarning)
-		{
-			if (!_getControlErrorMessageUsingThisJaspControl(control))
-				control->setHasWarning(false);
-		}
-	}
+void AnalysisForm::clearFormWarnings()
+{
+	_formWarnings.clear();
+	emit warningsChanged();
 }
 
-void AnalysisForm::formCompletedHandler()
+void AnalysisForm::setAnalysis(QVariant analysis)
 {
-	QTimer::singleShot(0, this, &AnalysisForm::_formCompletedHandler);
+	Analysis * analysisPointer = qobject_cast<Analysis *>(analysis.value<QObject *>());
+
+	if(_analysis == analysisPointer) return;
+
+	if(_analysis && analysisPointer)
+		throw std::runtime_error("An analysis of an analysisform was replaced by another analysis, this is decidedly NOT supported!");
+
+	_analysis = analysisPointer;
+
+	setAnalysisUp();
```

Comment:
I don't see the need to call that here.
Why just to set _analysis here and let formCompleted to call the setup? This makes more complicated than it is.... Now you need this extra boolean _formCompleted.

Reply:
Because maybe `setAnalysis` is only set/called *after* `formCompleted`.
I don't see how right now, but I prefer to safeguard the code for it like this.

And I would also like to get rid of all that 'formCompleted' stuff at some point because I have the feeling it makes things more rigid than necessary. Because everything has to be set then and won't be changed later if users change properties or whatever.